File: t/VB/MasterFormsEditor/UI/RenameOmrFieldsDlg.vb
' *************************************************************
' Copyright (c) 1991-2019 LEAD Technologies, Inc.              
' All Rights Reserved.                                         
' *************************************************************

Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports System.Windows.Forms
Imports System

Partial Public Class RenameOmrFieldsDlg
   Inherits Form
   Private _sampleName As String
   Private _mainForm As MainForm
   Private _omrFieldsCount As Integer
   Private _generatedNewNames As List(Of String)
   Private _alphabets As List(Of String)

   Public Sub New(mainForm As MainForm, omrFieldsCount As Integer)
      InitializeComponent()

      _omrFieldsCount = omrFieldsCount
      _mainForm = mainForm
      _generatedNewNames = New List(Of String)()
      _alphabets = GetAlphabets(omrFieldsCount)

      UpdateNames()

      Me.ActiveControl = _txtPrefix
   End Sub

   Public Shared Function GetAlphabets(fieldsCount As Integer) As List(Of String)
      Dim list As New List(Of String)()
      For i As Int16 = Convert.ToInt16("A"c) To Convert.ToInt16("Z"c)
         Dim c As Char = Convert.ToChar(i)
         list.Add(c.ToString())
         If list.Count = fieldsCount Then
            Return list
         End If
      Next

      For i1 As Int16 = Convert.ToInt16("A"c) To Convert.ToInt16("Z"c)
         Dim c1 As Char = Convert.ToChar(i1)
         For i2 As Int16 = Convert.ToInt16("A"c) To Convert.ToInt16("Z"c)
            Dim c2 As Char = Convert.ToChar(i2)
            list.Add(c1.ToString() + c2.ToString())
            If list.Count = fieldsCount Then
               Return list
            End If
         Next
      Next

      For i1 As Int16 = Convert.ToInt16("A"c) To Convert.ToInt16("Z"c)
         Dim c1 As Char = Convert.ToChar(i1)
         For i2 As Int16 = Convert.ToInt16("A"c) To Convert.ToInt16("Z"c)
            Dim c2 As Char = Convert.ToChar(i2)
            For i3 As Int16 = Convert.ToInt16("A"c) To Convert.ToInt16("Z"c)
               Dim c3 As Char = Convert.ToChar(i3)
               list.Add(c1.ToString() + c2.ToString() + c3.ToString())
               If list.Count = fieldsCount Then
                  Return list
               End If
            Next
         Next
      Next

      Return list
   End Function

   Private Sub UpdateNames()
      If _generatedNewNames IsNot Nothing Then
         _generatedNewNames.Clear()
      End If

      _sampleName = _txtPrefix.Text

      If _rbNum1.Checked Then
         _numStartsWith.Enabled = False
         _sampleName += "1"

         For i As Integer = 1 To _omrFieldsCount
            Dim fieldName As String = _txtPrefix.Text + i.ToString()
            _generatedNewNames.Add(fieldName)
         Next
      ElseIf _rbAlpha.Checked Then
         _numStartsWith.Enabled = False
         _sampleName += "A"

         For i As Integer = 0 To _omrFieldsCount - 1
            Dim fieldName As String = _txtPrefix.Text + _alphabets(i)

            _generatedNewNames.Add(fieldName)
         Next
      Else
         _numStartsWith.Enabled = True
         _sampleName += _numStartsWith.Value.ToString()
         For i As Integer = Decimal.ToInt32(_numStartsWith.Value) To Decimal.ToInt32(_numStartsWith.Value) + _omrFieldsCount
            Dim fieldName As String = _txtPrefix.Text + i.ToString()
            _generatedNewNames.Add(fieldName)
         Next
      End If

      _lblName.Text = _sampleName
   End Sub

   Private Sub _txtPrefix_TextChanged(sender As Object, e As EventArgs) Handles _txtPrefix.TextChanged
      UpdateNames()
   End Sub

   Private Sub _rb_CheckedChanged(sender As Object, e As EventArgs) Handles _rbStartsWith.CheckedChanged, _rbNum1.CheckedChanged, _rbAlpha.CheckedChanged
      UpdateNames()
   End Sub

   Private Sub _btnOK_Click(sender As Object, e As EventArgs) Handles _btnOK.Click
      _mainForm.ApplyRenameMultipleFields(_generatedNewNames)
      Me.Close()
   End Sub

   Private Sub _btnCancel_Click(sender As Object, e As EventArgs) Handles _btnCancel.Click
      Me.Close()
   End Sub

   Private Sub _numStartsWith_ValueChanged(sender As Object, e As EventArgs) Handles _numStartsWith.ValueChanged
      UpdateNames()
   End Sub
End Class

File: t/CS/MasterFormsEditor/UI/RenameOmrFieldsDlg.cs
// *************************************************************
// Copyright (c) 1991-2019 LEAD Technologies, Inc.              
// All Rights Reserved.                                         
// *************************************************************
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;

namespace CSMasterFormsEditor.UI
{
    public partial class RenameOmrFieldsDlg : Form
    {
        private string _sampleName;
        private MainForm _mainForm;
        private int _omrFieldsCount;
        private List<string> _generatedNewNames;
        private List<string> _alphabets;
        
        public RenameOmrFieldsDlg(MainForm mainForm, int omrFieldsCount)
        {
            InitializeComponent();

            _omrFieldsCount = omrFieldsCount;
            _mainForm = mainForm;
            _generatedNewNames = new List<string>();
            _alphabets = GetAlphabets(omrFieldsCount);

            UpdateNames();

            this.ActiveControl = _txtPrefix;
        }

        public static List<string> GetAlphabets(int fieldsCount)
        {
            List<string> list = new List<string>();

            for (char c = 'A'; c <= 'Z'; c++)
            {
                list.Add(c.ToString());
                if (list.Count == fieldsCount)
                    return list;
            }

            for (char c1 = 'A'; c1 <= 'Z'; c1++)
            {
                for (char c2 = 'A'; c2 <= 'Z'; c2++)
                {
                    list.Add(c1.ToString() + c2.ToString());
                    if (list.Count == fieldsCount)
                        return list;
                }
            }

            for (char c1 = 'A'; c1 <= 'Z'; c1++)
            {
                for (char c2 = 'A'; c2 <= 'Z'; c2++)
                {
                    for (char c3 = 'A'; c3 <= 'Z'; c3++)
                    {
                        list.Add(c1.ToString() + c2.ToString() + c3.ToString());
                        if (list.Count == fieldsCount)
                            return list;
                    }
                }
            }

            return list;
        }

        private void UpdateNames()
        {
            _generatedNewNames.Clear();
            _sampleName = _txtPrefix.Text;
                        
            if (_rbNum1.Checked)
            {
                _numStartsWith.Enabled = false;
                _sampleName += "1";
                
                for (int i = 1; i <= _omrFieldsCount; ++i)
                {
                    string fieldName = _txtPrefix.Text + i.ToString();
                    _generatedNewNames.Add(fieldName);
                }
            }
            else if (_rbAlpha.Checked)
            {
                _numStartsWith.Enabled = false;
                _sampleName += "A";

                for (int i = 0; i < _omrFieldsCount; ++i )
                {
                    string fieldName = _txtPrefix.Text + _alphabets[i];

                    _generatedNewNames.Add(fieldName);
                }
            }
            else
            {
                _numStartsWith.Enabled = true;
                _sampleName += _numStartsWith.Value.ToString();
                for (int i = decimal.ToInt32(_numStartsWith.Value); i <= decimal.ToInt32(_numStartsWith.Value) + _omrFieldsCount; ++i)
                {
                    string fieldName = _txtPrefix.Text + i.ToString();
                    _generatedNewNames.Add(fieldName);
                }
            }

            _lblName.Text = _sampleName;
        }

        private void _txtPrefix_TextChanged(object sender, EventArgs e)
        {
            UpdateNames();
        }

        private void _rb_CheckedChanged(object sender, EventArgs e)
        {
            UpdateNames();
        }

        private void _btnOK_Click(object sender, EventArgs e)
        {
            _mainForm.ApplyRenameMultipleFields(_generatedNewNames);
            this.Close();
        }

        private void _btnCancel_Click(object sender, EventArgs e)
        {
            this.Close();
        }

        private void _numStartsWith_ValueChanged(object sender, EventArgs e)
        {
            UpdateNames();
        }
    }
}
